source/common/request_info/dynamic_metadata_impl.cc

```diff
@@ -0,0 +1,53 @@
+#include "common/request_info/dynamic_metadata_impl.h"
+
+#include "envoy/common/exception.h"
+
+namespace Envoy {
+namespace RequestInfo {
+
+DynamicMetadataImpl::~DynamicMetadataImpl() {
+  for (auto& it : data_storage_) {
+    if (it.second.destructor_) {
+      it.second.destructor_(it.second.ptr_);
+    }
+  }
+  data_storage_.clear();
+}
+
+void DynamicMetadataImpl::setDataGeneric(absl::string_view data_name, size_t type_id, void* data,
+                                         void (*destructor)(void*)) {
+  if (data_storage_.find(data_name) != data_storage_.end()) {
+    (*destructor)(data);
+    throw EnvoyException("DynamicMetadata::setData<T> called twice with same name.");
```

Comment:
Should this just be an `ASSERT`?

Reply:
My inclination is against, but there may be issues I'm not thinking about.  

My argument would be that this can occur because of a configuration error (imagine two filters that are variants of each other that set the same output data, and someone configures them in the same filter chain), and I believe that throws result in cleanly handled configuration errors where asserts always crash.  But happy to take your guidance.